### PR TITLE
Add server core testing

### LINKED ISSUE #40804: Need to include Windows Server Core in our test suite
The image to use is available here https://github.com/dotnet/dotnet-buildtools-prereqs-docker/issues/316

There is no need to build here but we need to include this in our rolling build validation matrix as it's part of our supported platforms.

### PR DESCRIPTION
**Pull request by safern**
Fixes: https://github.com/dotnet/runtime/issues/40804

cc: @dotnet/runtime-infrastructure

### CODE REVIEW

**Comment by safern** [comment]
Let's see what CI results say.

**Comment by safern** [comment]
@MattGal I'm seeing failures like:

```
Executed on b00054H using docker image mcr.microsoft.com/dotnet-buildtools/prereqs:windowsservercore-1909-helix-amd64-20200521124620-7c805e9
running %HELIX_CORRELATION_PAYLOAD%\scripts\68c1891d04514baabf5ae0c927c1153e\execute.cmd in C:\h\w\9F6B097D\w\A2DB08D6\e max 900 seconds

Output:
Unable to pull image mcr.microsoft.com/dotnet-buildtools/prereqs:windowsservercore-1909-helix-amd64-20200521124620-7c805e9
Exit Code:-4
```

What could be causing this? I tried doing this locally and it worked fine:

```
docker run --rm -ti mcr.microsoft.com/dotnet-buildtools/prereqs:windowsservercore-1909-helix-amd64-20200521124620-7c805e9
```

**Comment by MattGal** [comment]
> @MattGal I'm seeing failures like:
> 
> ```
> Executed on b00054H using docker image mcr.microsoft.com/dotnet-buildtools/prereqs:windowsservercore-1909-helix-amd64-20200521124620-7c805e9
> running %HELIX_CORRELATION_PAYLOAD%\scripts\68c1891d04514baabf5ae0c927c1153e\execute.cmd in C:\h\w\9F6B097D\w\A2DB08D6\e max 900 seconds
> 
> Output:
> Unable to pull image mcr.microsoft.com/dotnet-buildtools/prereqs:windowsservercore-1909-helix-amd64-20200521124620-7c805e9
> Exit Code:-4
> ```
> 
> What could be causing this? I tried doing this locally and it worked fine:
> 
> ```
> docker run --rm -ti mcr.microsoft.com/dotnet-buildtools/prereqs:windowsservercore-1909-helix-amd64-20200521124620-7c805e9
> ```

Windows docker isn't quite as easy to use as it is on linux.  In this case you're using a Server RS5 queue (that's 1809) with a 19H1 (that's 1909) image.  At some point I think this has been addressed, but for 1909 you need to have at least 19H2.  I'd try windows.10.amd64.server20h1.open as the host queue.

**Comment by safern** [comment]
Ah good catch. I'll try that.

**Comment by ViktorHofer** [comment]
@safern is this PR ready to review?

**Comment by safern** [comment]
You can review it, but we still don't have the capacity we need to merge.

**Comment by safern** [comment]
This depends on: https://github.com/dotnet/dotnet-buildtools-prereqs-docker/pull/366 because the current image is 1909 and we don't have 1909 server pool on helix only 20H1 and we can't run the 1909 image there, so I'm adding a 20H1 image to be able to run the tests.

**Comment by danmosemsft** [comment]
Do you think we'll be able to get test results this week @safern ?

**Comment by safern** [comment]
> Do you think we'll be able to get test results this week @safern ?

@danmosemsft I just got the first set of results. Only 25 failures, nothing has jumped out yet, but still looking at them. I actually just opened one issue: https://github.com/dotnet/runtime/issues/42000

https://dev.azure.com/dnceng/public/_build/results?buildId=802835&view=ms.vss-test-web.build-test-results-tab

**Comment by danmosemsft** [comment]
Thanks, likely most or all of the failures will turn out to be tests that need capability checks modified. After all we used to run tests on Server Core in the past right? But it is good to check and get the run clean.

File: src/libraries/Common/tests/TestUtilities/System/PlatformDetection.Windows.cs [hunk 1]
```diff
@@ -30,6 +30,7 @@ public static partial class PlatformDetection
         public static bool IsNotWindows8x => !IsWindows8x;
         public static bool IsNotWindowsNanoServer => !IsWindowsNanoServer;
         public static bool IsNotWindowsServerCore => !IsWindowsServerCore;
+        public static bool IsNotWindowsNanoAndServerCore => IsNotWindowsNanoServer && IsNotWindowsServerCore;
```
Comment (danmosemsft) [inline src/libraries/Common/tests/TestUtilities/System/PlatformDetection.Windows.cs:33]: Nit, I think this should be `IsNotWindowsNanoNorServerCore` or `IsNeitherWindowsNanoNorServerCore` or `IsNotWindowsNanoAndIsNotServerCore`. It sounds like there's an OS named WindowsNanoAndServerCore, and we're not on that..

File: src/libraries/System.Diagnostics.Process/tests/ProcessStandardConsoleTests.cs [hunk 2]
```diff
@@ -15,6 +16,11 @@ public class ProcessStandardConsoleTests : ProcessTestBase
         [ConditionalFact(typeof(RemoteExecutor), nameof(RemoteExecutor.IsSupported))]
         public void TestChangesInConsoleEncoding()
         {
```
Comment (danmosemsft) [inline src/libraries/System.Diagnostics.Process/tests/ProcessStandardConsoleTests.cs:18]: I suggest just changing 
https://github.com/dotnet/runtime/blob/c0b87c70e42c98b03c3148a3a84eb6898ae3eb1c/src/libraries/System.Diagnostics.Process/tests/ProcessStandardConsoleTests.cs#L46
to
```c#
if (PlatformDetection.IsWindowsServerCore || PlatformDetection.IsWindowsNano)
```

I think your observation in #42000 that the test app is behaving differently to the test is interesting, but for the purposes of getting Server Core going, the above change just recognizes the test is behaving same as Nano.

**Review by danmoseley — APPROVED** [review]

File: src/libraries/System.Diagnostics.Process/tests/ProcessStandardConsoleTests.cs (hunk 2 above)
Comment (safern) [reply src/libraries/System.Diagnostics.Process/tests/ProcessStandardConsoleTests.cs:18]: Sounds reasonable. Will update to do that.

File: src/libraries/Common/tests/TestUtilities/System/PlatformDetection.Windows.cs (hunk 1 above)
Comment (safern) [reply src/libraries/Common/tests/TestUtilities/System/PlatformDetection.Windows.cs:33]: Sounds good, thanks for the feedback.

File: src/libraries/Common/tests/TestUtilities/System/PlatformDetection.cs [hunk 3]
```diff
@@ -226,17 +226,17 @@ private static bool GetIsInContainer()
             if (IsWindows)
             {
                 string key = @"HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control";
-                string value = "";
+                object value = null;
 
                 try
                 {
-                    value = (string)Registry.GetValue(key, "ContainerType", defaultValue: "");
+                    value = Registry.GetValue(key, "ContainerType", defaultValue: null);
```
Comment (eiriktsarpalis) [inline src/libraries/Common/tests/TestUtilities/System/PlatformDetection.cs:233]: Nit:

```suggestion
                    return Registry.GetValue(key, "ContainerType", defaultValue: null) is not null;
```

Not sure why we keep this in a try/catch block? Which type of exception is it compensating for?

Cf. https://github.com/dotnet/runtime/pull/42062/commits/5f845f6f1e1525f91127da6c6b1de3be1cae95fe#diff-9bc9aae45b52414dacb7f1ab20dfb587L62

Comment (safern) [reply src/libraries/Common/tests/TestUtilities/System/PlatformDetection.cs:233]: I'm not sure which type of exception it is compensating for. I'll update it.

**Comment by safern** [comment]
Failure is unrelated.

@directhex is this failure known?

**Pull request merged**